lib/saml/auth_fail_handler.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+module SAML
+  class AuthFailHandler
+    attr_accessor :message, :level, :context
+
+    CLICKED_DENY_MSG = 'Subject did not consent to attribute release'
+    TOO_LATE_MSG     = 'Current time is on or after NotOnOrAfter condition'
+    TOO_EARLY_MSG    = 'Current time is earlier than NotBefore condition'
+
+    def initialize(saml_response)
+      @saml_response = saml_response
+      @message = nil
+      @level   = nil
+      @context = nil
+    end
+
+    def known_error?
+      known_errors = [
+        :clicked_deny?,
+        :auth_too_late?,
+        :auth_too_early?
+      ]
+
+      known_errors.each do |known_error|
+        break if send(known_error)
+      end
+
+      !@message.nil? && !@level.nil?
+    end
+
+    def generic_error_message
+      message = <<-MESSAGE.strip_heredoc
+        SAML Login attempt failed! Reasons...
+          saml:    'valid?=#{@saml_response.is_valid?} errors=#{@saml_response.errors}'
+          user:    'valid?=#{@current_user&.valid?} errors=#{@current_user&.errors&.full_messages}'
```

Comment:
If `@saml_response.is_valid? == false`, then `@current_user` and `@session` will be `nil`, but we're safe because I'm using the safe navigator (`&.`) when printing the error.

If `@saml_response.is_valid? == true`, then they come from inside `sessions_controller#persist_session_and_user`.